### PR TITLE
GC tests: start creating closure table

### PR DESCRIPTION
**Pull request by osa1**
In #2573 (0f55c58) I omitted creating creating closure tables and
checking them after GC, implementing it now.

### CODE REVIEW

File: rts/motoko-rts-tests/src/gc.rs [hunk 1]
```diff
@@ -222,6 +254,23 @@ fn compute_reachable_objects(
...
+    assert_eq!(read_word(heap, offset), closure_table.len() as u32);
+    offset += WORD_SIZE;
+
+    for obj in closure_table.iter() {
+        let ptr = unskew_pointer(read_word(heap, offset));
+        offset += WORD_SIZE;
+
+        // Skip object header for idx
+        let idx_address = ptr as usize + WORD_SIZE;
+        let idx = read_word(heap, idx_address - heap.as_ptr() as usize);
```
Comment (osa1) [inline rts/motoko-rts-tests/src/gc.rs:270]: Forgot to compare `idx` with the expected one here, update coming.

Comment (osa1) [reply rts/motoko-rts-tests/src/gc.rs]: Done.

File: rts/motoko-rts-tests/src/gc.rs [hunk 2]
```diff
@@ -222,6 +254,26 @@ fn compute_reachable_objects(
...
+    assert_eq!(read_word(heap, offset), closure_table.len() as u32);
+    offset += WORD_SIZE;
+
+    for obj in closure_table.iter() {
+        let ptr = unskew_pointer(read_word(heap, offset));
+        offset += WORD_SIZE;
+
+        // Skip object header for idx
+        let idx_address = ptr as usize + size_of::<Array>().to_bytes().0 as usize;
+        let idx = get_scalar_value(read_word(heap, idx_address - heap.as_ptr() as usize));
```
Comment (crusso) [inline rts/motoko-rts-tests/src/gc.rs:270]: I don't understand what this line is doing, especially `idx_address - heap.as_ptr`?

Comment (osa1) [reply rts/motoko-rts-tests/src/gc.rs:273]: That is computing offset of index field of the object pointed by the table in `heap`. Pointers in heap are addresses, we need to convert those to offsets in `heap` to be able to access via `heap[...]`.

File: rts/motoko-rts-tests/src/gc/heap.rs [hunk 3]
```diff
@@ -289,13 +315,38 @@ fn create_dynamic_heap(
...
+        heap_offset += WORD_SIZE;
+
+        write_word(dynamic_heap, heap_offset, closure_table.len() as u32);
+        heap_offset += WORD_SIZE;
+
+        for idx in closure_table {
+            let idx_ptr = *object_addrs.get(idx).unwrap();
+            write_word(dynamic_heap, heap_offset, make_pointer(idx_ptr as u32));
+            heap_offset += WORD_SIZE;
+        }
```
Comment (crusso) [inline rts/motoko-rts-tests/src/gc/heap.rs:347]: This assumes there are no unused slots at the end of the closure table right? Ie. all the fields in the array will have sensible motoko values after this setup?

**Review by crusso — APPROVED** [review]

File: rts/motoko-rts-tests/src/gc.rs (hunk 2 above)
Comment (crusso) [reply rts/motoko-rts-tests/src/gc.rs:273]: Gotcha.

File: rts/motoko-rts-tests/src/gc/heap.rs (hunk 3 above)
Comment (osa1) [reply rts/motoko-rts-tests/src/gc/heap.rs:359]: Correct. I will generalize this as needed.

**Pull request merged**